import asyncio
import grpc
from typing import Optional, List, Dict, Any
from datetime import datetime
import logging

from .base_client import AsyncBaseClient
from .proto import simulator_pb2
from .proto import simulator_pb2_grpc
from .models import *
from .exceptions import *
from .utils import proto_to_dict

logger = logging.getLogger(__name__)


class AsyncSimulationClient(AsyncBaseClient):
    """
    –ê—Å–∏–Ω—Ö—Ä–æ–Ω–Ω—ã–π –∫–ª–∏–µ–Ω—Ç –¥–ª—è SimulationService.

    –†–∞–±–æ—Ç–∞–µ—Ç –Ω–∞ –ø–æ—Ä—Ç—É 50051 (–∏–ª–∏ –¥—Ä—É–≥–æ–º —É–∫–∞–∑–∞–Ω–Ω–æ–º –ø–æ—Ä—Ç—É).

    –ü—Ä–∏–º–µ—Ä –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏—è:
    ```python
    async with AsyncSimulationClient("localhost", 50051) as client:
        simulation = await client.create_simulation()
        await client.set_logist(simulation.simulation_id, "logist_123")
        results = await client.run_simulation(simulation.simulation_id)
    ```
    """

    def __init__(
        self,
        host: str = "localhost",
        port: int = 50051,  # üëà –ü–æ—Ä—Ç –¥–ª—è SimulationService
        max_retries: int = 3,
        timeout: float = 30.0,
        rate_limit: Optional[float] = None,
        enable_logging: bool = True,
    ):
        super().__init__(host, port, max_retries, timeout, rate_limit, enable_logging)
        self.stub = None

    async def connect(self):
        """–ü–æ–¥–∫–ª—é—á–∏—Ç—å—Å—è –∫ —Å–µ—Ä–≤–µ—Ä—É SimulationService."""
        try:
            self.channel = await self._create_channel()
            self.stub = simulator_pb2_grpc.SimulationServiceStub(self.channel)

            # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Å–æ–µ–¥–∏–Ω–µ–Ω–∏–µ
            if await self.ping():
                logger.info(
                    f"Connected to SimulationService at {self.host}:{self.port}"
                )
            else:
                raise ConnectionError(
                    f"Cannot connect to SimulationService at {self.host}:{self.port}"
                )

        except Exception as e:
            logger.error(f"Failed to connect to SimulationService: {e}")
            raise ConnectionError(f"Connection to SimulationService failed: {e}")

    async def close(self):
        """–ó–∞–∫—Ä—ã—Ç—å —Å–æ–µ–¥–∏–Ω–µ–Ω–∏–µ."""
        if self.channel:
            await self.channel.close()
            logger.info("Disconnected from SimulationService")

    async def ping(self) -> bool:
        """
        –ü—Ä–æ–≤–µ—Ä–∏—Ç—å –¥–æ—Å—Ç—É–ø–Ω–æ—Å—Ç—å SimulationService.

        Returns:
            bool: True –µ—Å–ª–∏ —Å–µ—Ä–≤–µ—Ä –¥–æ—Å—Ç—É–ø–µ–Ω
        """
        try:
            async with self._timeout_context(5.0):  # –ö–æ—Ä–æ—Ç–∫–∏–π —Ç–∞–π–º–∞—É—Ç –¥–ª—è ping
                await self._rate_limit()
                response = await self.stub.ping(simulator_pb2.PingRequest())
                return response.success
        except Exception as e:
            logger.warning(f"Ping to SimulationService failed: {e}")
            return False

    # ==================== –û—Å–Ω–æ–≤–Ω—ã–µ –æ–ø–µ—Ä–∞—Ü–∏–∏ —Å–∏–º—É–ª—è—Ü–∏–∏ ====================

    async def create_simulation(self) -> SimulationConfig:
        """
        –°–æ–∑–¥–∞—Ç—å –Ω–æ–≤—É—é —Å–∏–º—É–ª—è—Ü–∏—é.

        Returns:
            SimulationConfig: –ö–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏—è —Å–æ–∑–¥–∞–Ω–Ω–æ–π —Å–∏–º—É–ª—è—Ü–∏–∏
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                response = await self._with_retry(
                    self.stub.create_simulation, simulator_pb2.CreateSimulationRquest()
                )

                sim = response.simulation
                return SimulationConfig(
                    simulation_id=sim.simulation_id, capital=sim.capital
                )

        except grpc.RpcError as e:
            self._handle_grpc_error(e, "Create simulation")

    async def get_simulation(self, simulation_id: str) -> Dict[str, Any]:
        """
        –ü–æ–ª—É—á–∏—Ç—å –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –æ —Å–∏–º—É–ª—è—Ü–∏–∏.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏

        Returns:
            Dict: –ò–Ω—Ñ–æ—Ä–º–∞—Ü–∏—è –æ —Å–∏–º—É–ª—è—Ü–∏–∏
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                response = await self._with_retry(
                    self.stub.get_simulation,
                    simulator_pb2.GetSimulationRequest(simulation_id=simulation_id),
                )
                return proto_to_dict(response.simulation)

        except grpc.RpcError as e:
            self._handle_grpc_error(e, "Get simulation")

    async def run_simulation(self, simulation_id: str) -> SimulationResults:
        """
        –ó–∞–ø—É—Å—Ç–∏—Ç—å —Å–∏–º—É–ª—è—Ü–∏—é.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏

        Returns:
            SimulationResults: –†–µ–∑—É–ª—å—Ç–∞—Ç—ã —Å–∏–º—É–ª—è—Ü–∏–∏
        """
        try:
            async with self._timeout_context(
                self.timeout * 3
            ):  # –î–æ–ª—å—à–µ –¥–ª—è –∑–∞–ø—É—Å–∫–∞ —Å–∏–º—É–ª—è—Ü–∏–∏
                await self._rate_limit()
                response = await self._with_retry(
                    self.stub.run_simulation,
                    simulator_pb2.RunSimulationRequest(simulation_id=simulation_id),
                )

                sim = response.simulation
                return SimulationResults(
                    profit=sim.results.profit,
                    cost=sim.results.cost,
                    profitability=sim.results.profitability,
                    capital=sim.capital,
                    step=sim.step,
                    timestamp=datetime.fromisoformat(
                        response.timestamp.replace("Z", "+00:00")
                    ),
                )

        except grpc.RpcError as e:
            self._handle_grpc_error(e, "Run simulation")

    # ==================== –£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ –ª–æ–≥–∏—Å—Ç–∞–º–∏ ====================

    async def set_logist(self, simulation_id: str, worker_id: str) -> bool:
        """
        –ù–∞–∑–Ω–∞—á–∏—Ç—å –ª–æ–≥–∏—Å—Ç–∞ –¥–ª—è —Å–∏–º—É–ª—è—Ü–∏–∏.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            worker_id: ID —Ä–∞–±–æ—Ç–Ω–∏–∫–∞-–ª–æ–≥–∏—Å—Ç–∞

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.set_logist,
                    simulator_pb2.SetLogistRequest(
                        simulation_id=simulation_id, worker_id=worker_id
                    ),
                )
                logger.info(f"Set logist {worker_id} for simulation {simulation_id}")
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to set logist {worker_id}: {e}")
            return False

    # ==================== –£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ –ø–æ—Å—Ç–∞–≤—â–∏–∫–∞–º–∏ ====================

    async def add_supplier(
        self, simulation_id: str, supplier_id: str, is_backup: bool = False
    ) -> bool:
        """
        –î–æ–±–∞–≤–∏—Ç—å –ø–æ—Å—Ç–∞–≤—â–∏–∫–∞ –≤ —Å–∏–º—É–ª—è—Ü–∏—é.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            supplier_id: ID –ø–æ—Å—Ç–∞–≤—â–∏–∫–∞
            is_backup: –Ø–≤–ª—è–µ—Ç—Å—è –ª–∏ –∑–∞–ø–∞—Å–Ω—ã–º –ø–æ—Å—Ç–∞–≤—â–∏–∫–æ–º

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.add_supplier,
                    simulator_pb2.AddSupplierRequest(
                        simulation_id=simulation_id,
                        supplier_id=supplier_id,
                        is_backup=is_backup,
                    ),
                )
                logger.info(
                    f"Added supplier {supplier_id} to simulation {simulation_id}"
                )
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to add supplier {supplier_id}: {e}")
            return False

    async def delete_supplier(self, simulation_id: str, supplier_id: str) -> bool:
        """
        –£–¥–∞–ª–∏—Ç—å –ø–æ—Å—Ç–∞–≤—â–∏–∫–∞ –∏–∑ —Å–∏–º—É–ª—è—Ü–∏–∏.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            supplier_id: ID –ø–æ—Å—Ç–∞–≤—â–∏–∫–∞

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.delete_supplier,
                    simulator_pb2.DeleteSupplierRequest(
                        simulation_id=simulation_id, supplier_id=supplier_id
                    ),
                )
                logger.info(
                    f"Deleted supplier {supplier_id} from simulation {simulation_id}"
                )
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to delete supplier {supplier_id}: {e}")
            return False

    # ==================== –£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ —Å–∫–ª–∞–¥–æ–º ====================

    async def set_warehouse_worker(
        self, simulation_id: str, worker_id: str, warehouse_type: WarehouseType
    ) -> bool:
        """
        –ù–∞–∑–Ω–∞—á–∏—Ç—å —Ä–∞–±–æ—Ç–Ω–∏–∫–∞ –Ω–∞ —Å–∫–ª–∞–¥.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            worker_id: ID —Ä–∞–±–æ—Ç–Ω–∏–∫–∞
            warehouse_type: –¢–∏–ø —Å–∫–ª–∞–¥–∞

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.set_warehouse_inventory_worker,
                    simulator_pb2.SetWarehouseInventoryWorkerRequest(
                        simulation_id=simulation_id,
                        worker_id=worker_id,
                        warehouse_type=warehouse_type.value,
                    ),
                )
                logger.info(
                    f"Set worker {worker_id} on {warehouse_type.value} warehouse"
                )
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to set warehouse worker {worker_id}: {e}")
            return False

    async def increase_warehouse_size(
        self, simulation_id: str, warehouse_type: WarehouseType, size: int
    ) -> bool:
        """
        –£–≤–µ–ª–∏—á–∏—Ç—å —Ä–∞–∑–º–µ—Ä —Å–∫–ª–∞–¥–∞.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            warehouse_type: –¢–∏–ø —Å–∫–ª–∞–¥–∞
            size: –ù–∞ —Å–∫–æ–ª—å–∫–æ —É–≤–µ–ª–∏—á–∏—Ç—å

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.increase_warehouse_size,
                    simulator_pb2.IncreaseWarehouseSizeRequest(
                        simulation_id=simulation_id,
                        warehouse_type=warehouse_type.value,
                        size=size,
                    ),
                )
                logger.info(
                    f"Increased {warehouse_type.value} warehouse size by {size}"
                )
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to increase warehouse size: {e}")
            return False

    # ==================== –£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ —Ä–∞–±–æ—á–∏–º–∏ –º–µ—Å—Ç–∞–º–∏ ====================

    async def set_worker_on_workplace(
        self, simulation_id: str, worker_id: str, workplace_id: str
    ) -> bool:
        """
        –ù–∞–∑–Ω–∞—á–∏—Ç—å —Ä–∞–±–æ—Ç–Ω–∏–∫–∞ –Ω–∞ —Ä–∞–±–æ—á–µ–µ –º–µ—Å—Ç–æ.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            worker_id: ID —Ä–∞–±–æ—Ç–Ω–∏–∫–∞
            workplace_id: ID —Ä–∞–±–æ—á–µ–≥–æ –º–µ—Å—Ç–∞

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.set_worker_on_workerplace,
                    simulator_pb2.SetWorkerOnWorkerplaceRequest(
                        simulation_id=simulation_id,
                        worker_id=worker_id,
                        workplace_id=workplace_id,
                    ),
                )
                logger.info(f"Set worker {worker_id} on workplace {workplace_id}")
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to set worker on workplace: {e}")
            return False

    async def set_equipment_on_workplace(
        self, simulation_id: str, workplace_id: str, equipment_id: str
    ) -> bool:
        """
        –£—Å—Ç–∞–Ω–æ–≤–∏—Ç—å –æ–±–æ—Ä—É–¥–æ–≤–∞–Ω–∏–µ –Ω–∞ —Ä–∞–±–æ—á–µ–µ –º–µ—Å—Ç–æ.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            workplace_id: ID —Ä–∞–±–æ—á–µ–≥–æ –º–µ—Å—Ç–∞
            equipment_id: ID –æ–±–æ—Ä—É–¥–æ–≤–∞–Ω–∏—è

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.set_equipment_on_workplace,
                    simulator_pb2.SetEquipmentOnWorkplaceRequst(
                        simulation_id=simulation_id,
                        workplace_id=workplace_id,
                        equipment_id=equipment_id,
                    ),
                )
                logger.info(f"Set equipment {equipment_id} on workplace {workplace_id}")
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to set equipment on workplace: {e}")
            return False

    # ==================== –£–ø—Ä–∞–≤–ª–µ–Ω–∏–µ —Ç–µ–Ω–¥–µ—Ä–∞–º–∏ ====================

    async def add_tender(self, simulation_id: str, tender_id: str) -> bool:
        """
        –î–æ–±–∞–≤–∏—Ç—å —Ç–µ–Ω–¥–µ—Ä –≤ —Å–∏–º—É–ª—è—Ü–∏—é.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            tender_id: ID —Ç–µ–Ω–¥–µ—Ä–∞

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.add_tender,
                    simulator_pb2.AddTenderRequest(
                        simulation_id=simulation_id, tender_id=tender_id
                    ),
                )
                logger.info(f"Added tender {tender_id} to simulation {simulation_id}")
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to add tender {tender_id}: {e}")
            return False

    async def delete_tender(self, simulation_id: str, tender_id: str) -> bool:
        """
        –£–¥–∞–ª–∏—Ç—å —Ç–µ–Ω–¥–µ—Ä –∏–∑ —Å–∏–º—É–ª—è—Ü–∏–∏.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            tender_id: ID —Ç–µ–Ω–¥–µ—Ä–∞

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.delete_tender,
                    simulator_pb2.RemoveTenderRequest(
                        simulation_id=simulation_id, tender_id=tender_id
                    ),
                )
                logger.info(
                    f"Deleted tender {tender_id} from simulation {simulation_id}"
                )
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to delete tender {tender_id}: {e}")
            return False

    # ==================== –î–æ–ø–æ–ª–Ω–∏—Ç–µ–ª—å–Ω—ã–µ –Ω–∞—Å—Ç—Ä–æ–π–∫–∏ ====================

    async def set_dealing_with_defects(self, simulation_id: str, policy: str) -> bool:
        """
        –£—Å—Ç–∞–Ω–æ–≤–∏—Ç—å –ø–æ–ª–∏—Ç–∏–∫—É —Ä–∞–±–æ—Ç—ã —Å –±—Ä–∞–∫–æ–º.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            policy: –ü–æ–ª–∏—Ç–∏–∫–∞ —Ä–∞–±–æ—Ç—ã —Å –±—Ä–∞–∫–æ–º

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.set_dealing_with_defects,
                    simulator_pb2.SetDealingWithDefectsRequest(
                        simulation_id=simulation_id, dealing_with_defects=policy
                    ),
                )
                logger.info(
                    f"Set defects policy to {policy} for simulation {simulation_id}"
                )
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to set defects policy: {e}")
            return False

    async def set_certification(
        self, simulation_id: str, has_certification: bool
    ) -> bool:
        """
        –£—Å—Ç–∞–Ω–æ–≤–∏—Ç—å –Ω–∞–ª–∏—á–∏–µ —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ü–∏–∏.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            has_certification: –ï—Å—Ç—å –ª–∏ —Å–µ—Ä—Ç–∏—Ñ–∏–∫–∞—Ü–∏—è

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.set_has_certification,
                    simulator_pb2.SetHasCertificationRequest(
                        simulation_id=simulation_id, has_certification=has_certification
                    ),
                )
                status = "with" if has_certification else "without"
                logger.info(f"Set simulation {simulation_id} {status} certification")
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to set certification: {e}")
            return False

    async def add_production_improvement(
        self, simulation_id: str, improvement: str
    ) -> bool:
        """
        –î–æ–±–∞–≤–∏—Ç—å —É–ª—É—á—à–µ–Ω–∏–µ –ø—Ä–æ–∏–∑–≤–æ–¥—Å—Ç–≤–∞.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            improvement: –£–ª—É—á—à–µ–Ω–∏–µ –ø—Ä–æ–∏–∑–≤–æ–¥—Å—Ç–≤–∞

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.add_production_improvement,
                    simulator_pb2.AddProductionImprovementRequest(
                        simulation_id=simulation_id, production_improvement=improvement
                    ),
                )
                logger.info(
                    f"Added improvement {improvement} to simulation {simulation_id}"
                )
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to add production improvement: {e}")
            return False

    async def set_sales_strategy(self, simulation_id: str, strategy: str) -> bool:
        """
        –£—Å—Ç–∞–Ω–æ–≤–∏—Ç—å —Å—Ç—Ä–∞—Ç–µ–≥–∏—é –ø—Ä–æ–¥–∞–∂.

        Args:
            simulation_id: ID —Å–∏–º—É–ª—è—Ü–∏–∏
            strategy: –°—Ç—Ä–∞—Ç–µ–≥–∏—è –ø—Ä–æ–¥–∞–∂

        Returns:
            bool: True –µ—Å–ª–∏ —É—Å–ø–µ—à–Ω–æ
        """
        try:
            async with self._timeout_context():
                await self._rate_limit()
                await self._with_retry(
                    self.stub.set_sales_strategy,
                    simulator_pb2.SetSalesStrategyRequest(
                        simulation_id=simulation_id, sales_strategy=strategy
                    ),
                )
                logger.info(
                    f"Set sales strategy to {strategy} for simulation {simulation_id}"
                )
                return True

        except grpc.RpcError as e:
            logger.error(f"Failed to set sales strategy: {e}")
            return False
